.golangci.yaml

```diff
@@ -28,21 +28,25 @@ linters:
     - wastedassign
     - whitespace
     - protogetter
-    # TODO enable https://github.com/openfga/openfga/issues/598
-    # - bodyclose
-    # - contextcheck
-    # - exhaustive
-    # - errorlint
-    # - gochecknoglobals
-    # - gocritic
-    # - godot
-    # - gosec
-    # - noctx
-    # - inamedparam
-    # - perfsprint
-    # - prealloc
-    # - tparallel
+    - revive
+    - godot
+    - bodyclose
+    - gocritic
+    - godot
+    - gocyclo
+    #- gosec Too many different issues. Needs to be tackled in a separate PR.
+
 linters-settings:
+  gofmt:
+    simplify: true
+  staticcheck:
+    checks: [ "all" ]
+  gocyclo:
+    min-complexity: 58 # This is our baseline at the moment, but our goal is to take this down to 15.
```

Comment:
why 15? seems very arbitrary. Also, are we actually going to work on it? Otherwise this goal is wishful thinking 🤷‍♀️

Reply:
The concept of cyclomatic complexity was invented by Thomas J. McCabe in 1976 and based on his research:

## Ideal Cyclomatic Complexity any language
### Range Considerations:

- A **cyclomatic complexity between 1 and 10** is typically seen as good and indicates relatively simple code that should be easy to test and maintain.
- A score **above 10** might start to indicate more complex code where it might be beneficial to consider refactoring to simplify methods/functions, especially when it goes above 20. Code with complexity between 10 and 20 is still manageable, but caution should be used as complexity grows.
- A cyclomatic complexity **over 20** is often regarded as a sign of software that could be difficult to understand and maintain. Such complexity levels should prompt a review and potential refactoring, possibly breaking down complex functions into smaller, more manageable pieces.

--- 

@miparnisari 15 seems like a good average based on those ranges. I can remove the comment as most likely we won't work on it right now, but I'd keep track of the complexity so it doesn't go higher than what it currently is, otherwise the codebase will get extremely unmanageable.

Reply:
I think measuring and enforcing the cyclomatic complexity is well intentioned idea but it is imperfect measure of complexity and even then not a good measure of code _quality_. I'd much prefer our codebase to be readable and high-quality rather than adhering to some linter's rules. This would be a case of the tail wagging the dog.

Reply:
i'm okay with tracking the value. Our next step could be reducing the number to say 40 and see where we break this, this would be a good indication of which bits are complex

Reply:
@willvedd 
Cyclomatic complexity is **closely tied** with code quality but it should not be used in isolation. It's a metric used as a proxy to assess aspects of code quality, particularly maintainability and testability. 

### Testability
- **Higher Complexity, Lower Testability:** A higher cyclomatic complexity indicates more independent paths through the program, which typically means more test cases are needed to achieve adequate coverage. This can make comprehensive testing more challenging and time-consuming, potentially lowering the overall testability of the code.
- **Test Coverage Metrics:** Cyclomatic complexity directly influences the minimum number of test cases required for thorough testing. This is because each unique path ideally needs to be tested to ensure all logic branches are executed, which helps in detecting bugs or unexpected behavior in those branches.

### Maintainability

- **Complexity and Maintenance Efforts:** More complex code (i.e., code with a higher cyclomatic complexity) can be harder to understand and modify. Each decision point can introduce a potential error or require changes when modifications are made, increasing the maintenance burden.
- **Refactoring Triggers:** High complexity values often signal that a piece of code may need to be refactored into simpler, more manageable components. Refactoring can help reduce cyclomatic complexity by breaking down complex functions into simpler ones, making the code easier to maintain.

### Readability 

- **Indirect Impact:** While cyclomatic complexity does not measure readability directly, highly complex code can be more difficult to read and understand. Code that is hard to understand can be error-prone and may deter modifications by other developers due to the fear of introducing bugs.


--- 

Let's please consider the fact that right now inside this config the value is 58, which is incredibly high and lenient. It means that it's allowing functions that have over 1000 lines of code to exist. Ideally we take this down over time as @miparnisari suggests, e.g. to 40. This is mostly so we can start tracking another useful metric.

Reply:
Tangentially related: https://goreportcard.com/report/github.com/openfga/openfga

Why is this not complaining about `gocyclo`? 😆 
<img width="1203" alt="image" src="https://github.com/openfga/openfga/assets/5374887/77cf76e2-daf3-4220-bb57-58000977a193">